### ISSUE TITLE
False positive following `case None` in `match` statement (missed narrowing?)

### ISSUE THREAD

**Issue by mthuurne** [python/mypy#13046]
**Bug Report**

In `match` block with a `case None:`, mypy considers values captured by subsequent cases to still have `None` as a possible type.

**To Reproduce**

Check the following code using mypy:
```py
def f(obj: object) -> str | None:
    return str(obj) if obj else None

def g(obj: object) -> str:
    match f(obj):
        case None:
            return "-"
        case text:
            return text
```

**Expected Behavior**

No errors would be reported: a case that matches the literal `None` can be treated as matching `NoneType`.

**Actual Behavior**

```
match_nonetype.py:9: error: Incompatible return value type (got "Optional[str]", expected "str")  [return-value]
                return text
                       ^
```

**Your Environment**

- Python version used: 3.10.4
- Operating system and version: openSUSE Linux Tumbleweed
- Mypy version used: 0.961 (compiled: yes)
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files):
```
disallow_any_generics=True
disallow_incomplete_defs=True
disallow_untyped_defs=True
namespace_packages=True
no_implicit_optional=True
no_implicit_reexport=True
pretty=True
show_error_codes=True
strict_equality=True
warn_no_return=True
warn_redundant_casts=True
warn_return_any=True
warn_unreachable=True
warn_unused_configs=True
warn_unused_ignores=True
```

**Comment by mthuurne** [comment]
The problem can be worked around by assigning the result of the function call to a local variable:

```py
def f(obj: object) -> str | None:
    return str(obj) if obj else None

def g(obj: object) -> str:
    x = f(obj)
    match x:
        case None:
            return "-"
        case text:
            return text
```

So this might be a variation of #12998.

**Comment by dset0x** [comment]
For what it's worth, it appears that the following is also currently accepted by mypy:
```
    match _ = f(obj):
        case None:
            return "-"
        case text:
            return text
```

**Comment by erictraut** [comment]
This is a duplicate of https://github.com/python/mypy/issues/12998.
